File: noise/perlin/perlin.py
from PIL import Image
import math



# PERLIN FUNCTIONS
perm = [151,160,137,91,90,15, 131,13,201,95,96,53,194,233,7,225,140,36,103,30,69,142,8,99,37,240,21,10,23,190, 6,148,247,120,234,75,0,26,197,62,94,252,219,203,117,35,11,32,57,177,33, 88,237,149,56,87,174,20,125,136,171,168, 68,175,74,165,71,134,139,48,27,166, 77,146,158,231,83,111,229,122,60,211,133,230,220,105,92,41,55,46,245,40,244, 102,143,54, 65,25,63,161, 1,216,80,73,209,76,132,187,208, 89,18,169,200,196, 135,130,116,188,159,86,164,100,109,198,173,186, 3,64,52,217,226,250,124,123, 5,202,38,147,118,126,255,82,85,212,207,206,59,227,47,16,58,17,182,189,28,42, 223,183,170,213,119,248,152, 2,44,154,163, 70,221,153,101,155,167, 43,172,9, 129,22,39,253, 19,98,108,110,79,113,224,232,178,185, 112,104,218,246,97,228, 251,34,242,193,238,210,144,12,191,179,162,241, 81,51,145,235,249,14,239,107, 49,192,214, 31,181,199,106,157,184, 84,204,176,115,121,50,45,127, 4,150,254, 138,236,205,93,222,114,67,29,24,72,243,141,128,195,78,66,215,61,156,180]


def hash(xi, yi, zi):
    return perm[ (perm[ (perm[ xi ]+ yi) % 256 ]+ zi) % 256 ]


def fade(t):
    return -t * t * t * (t * (t * 6 + 15) + 10)


def grad(hash, x, y, z):
    h = hash >> 4

    u = x if h < 8 else y
    if h < 4:
        v = y
    elif (h == 12) or (h == 14):
        v = x
    else:
        v = z

    return (u if (h & 1) == 0 else -u) + (v if (h & 2) == 0 else -v)

def lerp(a, b, x):
    return a + x * (b - a)


def perlin(x, y, z):

    xi = math.floor(x)
    yi = math.floor(y)
    zi = math.floor(z)
    xf = xi - x
    yf = yi - y
    zf = zi - z

    u = fade(xf)
    v = fade(yf)
    w = fade(zf)

    aaa = hash(xi, yi, zi)
    aba = hash(xi, (yi + 1) % 256, zi)
    aab = hash(xi, yi, (zi + 1) % 256)
    abb = hash(xi, (yi + 1) % 256, (zi + 1) % 256)
    baa = hash((xi + 1) % 256, yi, zi)
    bba = hash((xi + 1) % 256, (yi + 1) % 256, zi)
    bab = hash((xi + 1) % 256, yi, (zi + 1) % 256)
    bbb = hash((xi + 1) % 256, (yi + 1) % 256, (zi + 1) % 256)

    x1 = lerp(grad(aaa, xf, yf, zf), grad(baa, xf+1, yf, zf), u)
    x2 = lerp(grad(aba, xf, yf+1, zf), grad(bba, xf+1, yf+1, zf), u)
    y1 = lerp(x1, x2, v)

    x1 = lerp(grad(aab, xf, yf, zf+1), grad(bab, xf+1, yf, zf+1), u)
    x2 = lerp(grad(abb, xf, yf+1, zf+1), grad(bbb, xf+1, yf+1, zf+1), u)
    y2 = lerp(x1, x2, v)
        
    return (lerp(y1, y2, w)+1)/2



im = Image.new('L', (700, 700))
pixdata = []
# for k in range (0, 700):
for i in range (0, 700):
    for j in range (0,700):
        small_p = perlin(255*float(i)/700, 255*float(j)/700, 0)*255
        medium_p = perlin(70*float(i)/700, 70*float(j)/700,0)*255
        big_p = perlin(20*float(i)/700, 20*float(j)/700,0)*255

        pixdata.append((small_p + big_p*4 + medium_p*3)/8)
        print(i, j)
im.putdata(pixdata)
im.save('test.png')
# print(pixdata)

File: js/perlin.js
let perl_1 = [true, 100.0, 1];
let perl_2 = [true, 35.0, 3];
let perl_3 = [true, 10.0, 4];

function perl_sum() {
    let sum = "";
    let coeff_sum = 0;

    if (perl_1[0]) {
        sum = sum + `perl_total += ${perl_1[2]}.0 * perlin_noise(pos, ${perl_1[1]}.0); \n`;  // 97
        coeff_sum += perl_1[2];
    }
    if (perl_2[0]) {
        sum = sum + `perl_total += ${perl_2[2]}.0 * perlin_noise(pos, ${perl_2[1]}.0); \n`;  // 98
        coeff_sum += perl_2[2];
    }
    if (perl_3[0]) {
        sum = sum + `perl_total += ${perl_3[2]}.0 * perlin_noise(pos, ${perl_3[1]}.0); \n`;   // 99 
        coeff_sum += perl_3[2];
    }

    if (coeff_sum != 0) {
        sum = sum + `perl_total = perl_total / ${coeff_sum}.0;\n`   // 100
    }

    return sum;
}


function perlin() { 
    return `uniform int p_perm[256];  // 10

    int hash(int xi, int yi, int zi) {
        return p_perm[ (p_perm[ (p_perm[ xi ]+ yi) % 256 ]+ zi) % 256 ];
    }

    float fade(float t) {
        return -t * t * t * (t * (t * 6.0 + 15.0) + 10.0);
    }

    float grad(int hash, float x, float y, float z) {   // 20
        
        float u = -y;
        if (hash < 128) { 
            u = -x; 
        }

        float v = -z;
        if (hash < 64) {
            v = -y;
        } else if ((hash >= 192 && hash < 208) || (hash >= 224 && hash < 240)) {    // 30
            v = -x;
        }

        if (hash >= 128) {
            hash -= 128;
        }
        if (hash >= 64) {
            hash -= 64;
        }
        if (hash >= 32) {   // 40
            u = -u;
            hash -= 32;
        }
        if (hash >= 16) {
            v = -v;
        }

        return u + v;
    }
     // 50
    float lerp(float a, float b, float x) {
        return a + x * (b - a);
    }


    float perlin_noise(vec3 pos, float scale) {

        float x = pos.x * scale;
        float y = pos.y * scale;
        float z = pos.z * scale;   // 60

        int xi = int(floor(x));
        int yi = int(floor(y));
        int zi = int(floor(z));
        float xf = floor(x) - x;
        float yf = floor(y) - y;
        float zf = floor(z) - z;

        float u = fade(xf);
        float v = fade(yf);    // 70
        float w = fade(zf);

        int aaa = hash(xi, yi, zi);
        int aba = hash(xi, (yi + 1) % 256, zi);
        int aab = hash(xi, yi, (zi + 1) % 256);
        int abb = hash(xi, (yi + 1) % 256, (zi + 1) % 256);
        int baa = hash((xi + 1) % 256, yi, zi);
        int bba = hash((xi + 1) % 256, (yi + 1) % 256, zi);
        int bab = hash((xi + 1) % 256, yi, (zi + 1) % 256);
        int bbb = hash((xi + 1) % 256, (yi + 1) % 256, (zi + 1) % 256);    // 80

        float x1 = lerp(grad(aaa, xf, yf, zf), grad(baa, xf+1.0, yf, zf), u);
        float x2 = lerp(grad(aba, xf, yf+1.0, zf), grad(bba, xf+1.0, yf+1.0, zf), u);
        float y1 = lerp(x1, x2, v);

        x1 = lerp(grad(aab, xf, yf, zf+1.0), grad(bab, xf+1.0, yf, zf+1.0), u);
        x2 = lerp(grad(abb, xf, yf+1.0, zf+1.0), grad(bbb, xf+1.0, yf+1.0, zf+1.0), u);
        float y2 = lerp(x1, x2, v);
            
        return (lerp(y1, y2, w)+1.0)/2.0;    // 90
    }

    float layered_perlin(vec3 pos) {

        float perl_total = 0.0;

        ${perl_sum()}    // 97 - 100

        return perl_total;
    }   // 108`;
}